---
title: "BNT_BNT clinical samples proteomics"
output:
  pdf_document: default
  html_document:
    df_print: paged
    toc: TRUE
---
### Proteomics using MS-DAP
The samples were all prepared and analysed within the same batch and then analysed with DIA-NN.

```{r, echo = F, message=FALSE, warning = F}
 opts = list(comment = FALSE)
library(devtools)
#devtools::install_github("https://github.com/vdemichev/diann-rpackage")
library(diann)
#devtools::install_github("ftwkoopmans/msdap", upgrade = "never") # don't update dependencies if not needed
library(msdap)
library(readxl)
library(tidyverse)
library(ggplot2)
library(ggrepel)
library(reshape2)
```

```{r get_swissprot_fasta, echo = F}
# proteinIds <- read.delim("~/Downloads/muelledm@20210812+1504_50-0075/report.tsv") %>%
#   select(Protein.Ids) %>%
#   distinct()
# IDs = sapply(., function(x) unlist(strsplit(x, ";")))
# write.csv(IDs, "../human_Swissport_IDs_clinical_vaccine_COVID.csv")
```

```{r preprocess, echo = F, eval = F}

report <- read.delim("../report.tsv") 
report <- report %>% 
  mutate(Run = sapply(report$Run, function(x) strsplit(x, ".", fixed = T)[[1]][1])) %>% 
  mutate(Precursor.Quantity = log2(Precursor.Quantity+0.00001))
meta <- read_excel("../20210909_50-0075_InnateVaccineResponse.xlsx") %>% 
  na.omit() %>% 
  rename(Run = sample_id) %>% 
  mutate(group = paste(vacc_group, visit, sep = "_")) #paste(vacc_group, visit, sep = "_")

#get report index
meta_run =  data.frame(Run = meta$Run)
report_run = data.frame(Run = report$Run)

report_modified <- left_join(meta_run, report)

#openxlsx::write.xlsx(report_modified, file = "../report_IVR.xlsx", overwrite = T)
write.table(report_modified, file='../report_IVR.tsv', quote=FALSE, sep='\t', row.names = F)

```


```{r, echo = F, eval = F}

write_template_for_sample_metadata(dataset, "../sample_metadata.xlsx", overwrite = T)
orig_meta <- read_excel("../20210909_50-0075_InnateVaccineResponse.xlsx")
orig_meta <- orig_meta %>%
  mutate(sample_id = sapply(orig_meta$sample_id, function(x) paste0(x, ".rawIdx")))
  #%>% na.omit()

IVR_meta <- openxlsx::read.xlsx("../sample_metadata.xlsx")
IVR_meta <- IVR_meta %>%
  rename(plate_row = x3) %>%
  mutate(Run = sapply(IVR_meta$sample_id, function(x) strsplit(x, ".", fixed = T)[[1]][1])) %>%
  inner_join(., meta, by = "Run") %>%
  select(!c(x1, x2, x4, group.x)) %>%
  rename(group = group.y) %>%
  select(!Run)
openxlsx::write.xlsx(IVR_meta, file = "../sample_metadata.xlsx", overwrite = T)
#write.csv(IVR_meta, "../sample_metadata.xlsx", row.names = F)

#filename <- "/home/parnika/Documents/Projects/COV19_vaccine_proteomics/sample_metadata.xlsx"

dataset = import_dataset_diann(filename = "../report_IVR.tsv")
dataset = import_fasta(dataset, files = "../20200406_Human-isoform-uniprot-proteome-3AUP000005640.fasta")
#dataset = import_fasta(dataset, files = "../human_proteome_IVR_covvac.fasta")
dataset = msdap::import_sample_metadata(dataset, filename = "../sample_metadata_modified.xlsx")
```


## Volcano plots
```{r in-house to select diff abund proteins from msdap results, fig.dim=c(10,10), message=FALSE, warning = F}

volcano_plot <- function(dataset, contrast_list, algorithm)
{
  dataset$de_proteins %>% 
    select(protein_id, pvalue, qvalue, contrast, foldchange.log2, algo_de) %>% 
    #mutate(contrast = str_split(contrast, ": ", simplify = T)[,2]) %>% 
    left_join(., dataset$proteins) %>% 
    select(-fasta_headers, -accessions) %>% 
    rename(geneID = gene_symbols_or_id) %>% 
    filter(algo_de %in% algorithm) -> de 
  
  de %>% 
    rename(fc = starts_with("foldchange")) %>% 
    mutate(logPval = -log10(pvalue+(1e-8))) %>% 
    mutate(logQval = -log10(qvalue+(1e-8))) %>% 
    mutate(col = case_when(
      logPval >= 1.3 & abs(fc) >= 0.2 ~ "deepskyblue4",
      logPval < 1.3 & abs(fc) >= 0.2 ~ "firebrick",
      logPval >= 1.3 & abs(fc) < 0.2 ~ "darkolivegreen",
      TRUE ~ "gray"
    )) %>% 
    mutate(label = case_when(
      logPval < 1.3 & abs(fc) < 0.2 ~ "",
      TRUE ~ geneID
    )) %>% 
    mutate(shape = case_when(
      logQval >= 1.3 & abs(fc) >= 0.2 ~ 8,
      TRUE ~ 16,
      )) -> df
  
  for(i in algorithm)
    print(
      df %>% 
        filter(algo_de == i)  %>% 
      
        ggplot(aes(x = fc, y = logPval, colour = col, label = geneID, shape = shape)) +
        geom_point() +
        scale_colour_identity() +
        scale_shape_identity() +
        theme_bw() +
        geom_hline(yintercept = 1.3, linetype = "dashed", colour = "gray84") +
        geom_vline(xintercept = 0.2, linetype = "dashed", colour = "gray84") +
        geom_vline(xintercept = -0.2, linetype = "dashed", colour = "gray84") +
        facet_wrap(~ contrast, scales = "free", ncol = 2) +
        geom_text_repel(size = 2.2) +
        xlab("log2 fold change") +
        ylab("-log10 p-value") +
        ggtitle(i)
    )
  # ggsave(paste(strsplit(contrast_list[[1]][1], split = "visit")[[1]][1],
  #              "volcano.pdf", sep = ""))
  #return(p1)
}


dataset = import_dataset_diann(filename = "../report_IVR.tsv")
dataset = import_fasta(dataset, 
                       files = "../20200406_Human-isoform-uniprot-proteome-3AUP000005640.fasta")
dataset = import_sample_metadata(dataset, 
                                 filename = "../sample_metadata_modified.xlsx")

# contrast_list <- list(c("AZD_BNT_visit_0", "AZD_BNT_visit_1"),
#                        c("AZD_BNT_visit_0", "AZD_BNT_visit_4"),
#                        c("AZD_BNT_visit_0", "AZD_BNT_visit_5"),
#                        c("AZD_BNT_visit_5", "AZD_BNT_visit_6"),
#                        c("AZD_BNT_visit_5", "AZD_BNT_visit_8"))

contrast_list <- list(c("BNT_BNT_visit_0", "BNT_BNT_visit_1"),
                       c("BNT_BNT_visit_0", "BNT_BNT_visit_4"),
                       c("BNT_BNT_visit_0", "BNT_BNT_visit_5"),
                       c("BNT_BNT_visit_5", "BNT_BNT_visit_6"),
                       c("BNT_BNT_visit_5", "BNT_BNT_visit_8"))
algorithm <- c("ebayes", "deqms", "msempire") #"ebayes", "deqms", "msempire"


dataset = setup_contrasts(dataset, contrast_list = contrast_list)
  
dataset = analysis_quickstart(
  dataset,
  filter_min_detect = 2,         
  filter_min_quant = 3,         
  filter_fraction_detect = 0.5, 
  filter_fraction_quant = 0.5,  
  filter_by_contrast = T,     
  norm_algorithm = c("vsn", "modebetween"),
  dea_algorithm = algorithm, 
  dea_qvalue_threshold = 0.1,                      
  dea_log2foldchange_threshold = 0.2,                
  diffdetect_min_samples_observed = 1,
  pca_sample_labels = "auto",
  output_qc_report = F,                          
  output_abundance_tables = TRUE,                   
  output_dir = "msdap_results",                    
  output_within_timestamped_subdirectory = F,
  dump_all_data = F
  )
  
  
volcano_plot(dataset, contrast_list, algorithm)


```


## Boxplot to see protein abundance trend
```{r boxplots, results = "asis"}

boxplot_function <- function(dataset, contrast_list)
{
  contrast_unlist <- contrast_list %>%
    unlist() %>%
    unique()

  genelist <- dataset$de_proteins %>%
    filter(pvalue <= 0.05 & foldchange.log2 >= 0.2) %>%
    left_join(., dataset$proteins) %>%
    select(-fasta_headers, -accessions) %>% 
    group_by(gene_symbols_or_id) %>% 
    mutate(sigannot = case_when(
      qvalue <= 0.05 & foldchange.log2 >= 0.2 ~ "_p.adj",
      TRUE ~ "",
    )) %>% 
    select(gene_symbols_or_id, sigannot)
  
  unique_genes <- unique(genelist$gene_symbols_or_id)
  for(i in unique_genes)
  {
    pos = which(genelist$gene_symbols_or_id == i)
    if(any(genelist$sigannot[pos] == "_p.adj"))
      genelist[pos,3] <- paste(i, "_p.adj", sep = "")
    else{
      genelist[pos,3] <- i
    }
  }
  
  genelist <- genelist %>% 
    rename(gene_sigannot = "...3") %>% 
    select(gene_symbols_or_id, gene_sigannot) %>% 
    distinct()

  sample_id <- dataset$samples %>%
    filter(exclude == FALSE) %>% 
    mutate(sample_id = str_replace(sample_id, "-", "\\.")) %>%
    mutate(sample_used = paste(group, sample_id, sep = "_")) %>%
    filter(group %in% contrast_unlist)

  prot_used <- read.delim("msdap_results/results/protein_abundance__input data as-is.tsv") %>%
    setNames(str_replace(colnames(.), "X", "")) %>%
    select(gene_symbols_or_id,
           ends_with(sample_id$sample_id)) %>%
    filter(gene_symbols_or_id %in% genelist$gene_symbols_or_id) %>% 
    left_join(genelist)

  # because duplicate rownames are not allowed:
  .rowNamesDF(prot_used, make.names = T) <- prot_used$gene_sigannot

  prot_used_long <- prot_used %>%
    select(-gene_symbols_or_id, -gene_sigannot) %>% 
    t() %>%
    as.data.frame()
    
    prot_used_long$group = sapply(rownames(prot_used_long), 
                                  function(x) sample_id$group[grep(pattern = x, 
                                                                         sample_id$sample_id)])
    
 

  prot_long_melt <- melt(prot_used_long, id = "group") %>%
    rename(gene = "variable",
         abundance = "value") %>% 
    na.omit() #%>%
    #filter(gene %in% c("SAA1.SAA2", "C9", "LBP", "SERPINA3", "APOA4", "LRG1", "ITIH4"))
 
    # boxplot to show increase and decrease of protein abundances
  unique_genes_gg <- unique(prot_long_melt$gene)
  for(i in unique_genes_gg)
  {
    cat("  \n###", i, "  \n")
    print(
      prot_long_melt %>% 
        filter(gene == i) %>% 
        ggplot(aes(x = group, y = abundance, colour = group)) +
          geom_jitter() +
          geom_boxplot(alpha = 0.2) +
          #facet_wrap(~ i, scales = "free", ncol = 1) +
          theme_bw() +
          ylab("log2 (protein abundance)") +
          ggtitle(i) +
          theme(axis.ticks.x = element_blank(),
              axis.text.x = element_blank())
        )
    cat("  \n")
  }

  #ggsave(plot = p2, "AZD_BNT_boxplot.pdf", units = "cm", width = 30, height = 10)
  #return(prot_long_melt)
}

boxplot_function(dataset, contrast_list)

```














